Piecewise expr must be Basic

smichr opened the pull request:
## Summary of changes

* changes to ExprCondPair
  * expr must be Basic
  * cond must be Boolean or must be able to be reduced to Boolean
  * a cond containing Piecewise will be folded
  * a cond that is a Piecewise (perhaps after folding) will be rewritten as ITE with canonical Relationals
  * a TypeError is raised if the cond is not a Boolean
 * is is used instead of == since ExprCondPair args are Boolean now
* `piecewise_fold` will convert ITE conditions to nnf and simplified if possible
* ITE/And/Or take only Boolean args e.g. `And(x, y)` is okay, `And(2, x + 1, x**2)` is not.
* ITE now pretty prints like Piecewise
* BooleanFunction has a binary_symbols property which returns a set of symbols which can have a boolean/binary value
* as_Basic utility function to return the Basic version of an expression or raise an error if there is none.
* Piecewise._eval_subs is updated to reflect that Piecewise does not contain bool
* `piecewise_fold` now folds the conditions and if the condition (before or after folding) contains Boolean expressions, then it is converted to an unsimplified BooleanFunction.
```
>>> Piecewise((1,Piecewise((x>1,x<3),(x>2,True))))
Piecewise((1, ITE(x < 3, x > 1, x > 2)))
>>> piecewise_fold(_)
Piecewise((1, ((x >= 3) | (x > 1)) & ((x > 2) | (x < 3))))
>>> simplify(_)
Piecewise((1, x > 1))
```
* `as_Boolean` is added to boolalg which behaves like as_Basic
* Boolean gets default `as_set` and `to_nnf` methods
* args of ITE, And and Or must be Boolean (for ITE this required adding a `__new__` definition
* simplify_logic now does simplification of variables before checking if the expression is a BooleanFunction since this may change the input function's type
* Piecewise.rewrite(ITE)  has been added
* conditionals appearing in a solution of a Piecewise function are simplified
* differentiation of Boolean is simply unevaluated for expressions involving relationals in the symbol for which the derivative is requested unless the symbol is a binary symbol
```
>>> (x<1).diff(x)
Derivative(x < 1, x)
>>> And(x<1, x).diff(x)  # result does not depend on whether x is 0 or 1
0
>>> And(x<=1, x).diff(x)  #in this case it does: x =0 gives False while x=1 gives True
1
```
* `func_name` recognizes "<class 'foo'>" when trying to identify `foo`
* PEP8 edits of periodicity tests
* solve and checksol more carefully handle Eq/Ne expressions which have one arg that is True or False; in this case, the other arg must also be a Boolean
* interval solutions from solve are now disabled in Piecewise solving -- when solving for x one should not have x in the solution

(Much of this is split out of #12587)
closes #13223
closes #13255 on which it was built
fixes #13347
fixes #13349 
fixes #13370 
addressed #13350 but doesn't handle case of non-binary symbols; close if master doesn't handle Relationals like `(x**2 < 1).diff(x)`

- [x] check coverage

**REMOVED**
* careful simplification of BooleanFunction is added for univariate expressions
* simplification of complex univariate Relationals has been implemented through the use of sets provided that the involved expressions do not have periodic behavior.
* Relational._eval_simplify has been modified to use the newly added `simplify_relationships` which will try to simplify boolean functions via sets (with care not to add +/-oo when not necessary)

smichr commented:
ping @asmeurer -- just double checking that this is the intention (to have nothing but sympified args in objects).

asmeurer commented:
I think even Tuple in a Piecewise should be avoided, at least when dealing with expressions (but maybe not an error, I'm not sure). If the expression arg isn't an Expr it doesn't make mathematical sense to combine it with another expression.  `Piecewise(((1, 2), x < 0)) + 1` doesn't make sense, but `(Piecewise((1, x < 0)) + 1, Piecewise((2, x < 0)) + 1)` does.

smichr commented:
Ok. We should think of a Piecewise EXPRESSION not a Piecewise container or a convenient alternative to a nested ITE.

asmeurer commented:
Probably it could be (ab)used for other functions, but all the methods and simplification functions will assume that things are expressions. For booleans you definitely want to use ITE, as it knows about booleans and has the appropriate boolean methods (like to_nnf). For containers, probably if you are doing this it means you want to be using sets, which have their own ways of representing conditionals (like ConditionSet).

asmeurer commented:
(by the way I'm +1 to this)

smichr commented:
OK, let's not commit this yet. There is an XFAILed test that has a condition that has a Piecewise something like:

```
cond = Eq(Piecewise((cos(x), x<0), (a, True)), 0)
```
This could be made into `Piecewise((Eq(cos(x), 0), x<0),(Eq(a, 0), True))` but then the arg is not an Expr, it's a Boolean. So perhaps this is where ITE could be used to write this as `cond = ITE(x<0, Eq(cos(x), 0), Eq(a, 2)).to_nnf()`:

```
And(Or(Eq(a,2),x<0),Or(Eq(cos(x),0),x>=0))
```

asmeurer commented:
What's wrong with `Eq(Piecewise((cos(x), x<0), (a, True)), 0)`?

smichr commented:
> What's wrong with ...

Some of the problems that integration of Piecewise has (cf #11045) is that the code to check the domain represented by a condition doesn't fully anticipate what forms the cond part of the ExprCond pair will have. So you end up with AttributeErrors.  So one idea is to make `piecewise_fold` turn piecewise conditions into a Boolean expressions.

asmeurer commented:
Oh, I see, you have a Piecewise that has another Piecewise in a condition. Is it possible to denest that somehow?

asmeurer commented:
So say you have `Piecewise((f(x), Eq(Piecewise((cos(x), x<0), (a, True)), 0)), (g(x), True))`

```
⎧          ⎧cos(x)  for x < 0
⎪f(x)  for ⎨                  = 0
⎨          ⎩  a     otherwise
⎪
⎩g(x)          otherwise
```

Isn't this equivalent to `Piecewise((f(x), And(cos(x), x<0)), (f(x), Eq(a, 0)), (g(x), True))`

```
⎧f(x)  for (cos(x) ∨ a = 0) ∧ (a = 0 ∨ x < 0)
⎨
⎩g(x)                otherwise
```

which it looks like it is already simplifying in the same way ITE.to_nnf would:

```
>>> ITE(x < 0, Eq(cos(x), 0), Eq(a, 0)).to_nnf()
(Eq(a, 0) | (x < 0)) & ((x >= 0) | Eq(cos(x), 0))
```

But I understand where you're coming from now. I agree it makes more sense to put a boolean inside a Piecewise. Maybe we should do that. My only worry is that a lot of Piecewise code is written assuming the `expr` parts are `Expr` objects. 

`ITE` could be used but it would need a lot of cleaning up to make it more user friendly (better printing, support for multiple ITEs without nesting), better simplification in terms of higher level logic functions than just And and Or). I remember discussing this on another issue, but when the expressions themselves are booleans, you can simplify them with the conditions, which ITE already knows how to do.

smichr commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -546,6 +563,17 @@ def piecewise_fold(expr):
+            if c.has(Piecewise):
+                c = piecewise_fold(c)
+                if isinstance(c, Piecewise):
+                    # Piecewise((c1, k1), (c2, k2)) ->
```
ping @jksuom , can you give these few lines a once over, please? I am rewriting a Piecewise condition as an Or or Ands, e.g. `Piecewise((x < 1, y < 0), (x > 2, y < 1))` -> `(x < 1) & (y < 0) | (x > 2) & (y < 1) & ~(y < 0)`.

smichr replied:
Perhaps this should become a method of Piecewise so one could write a Piecewise of conditions and then convert it into a Boolean. Perhaps `Piecewise.as_boolean()` or `to_nnf` as for `ITE`?

smichr replied:
But I think this is the `dnf` form, so the method would be `to_dnf`.

smichr commented:
It would be good to have someone take a look at the `to_dnf` method.

smichr commented:
@asmeurer , what do you think about the `to_dnf` method? Should I call it `as_condition` instead? or something else? Or perhaps I could add an ITE rewrite method -- ITE has a PIecewise rewrite method -- and then use the `to_dnf` method from there.

smichr commented:
If #13255 is added, rewriting a Piecewise with no default condition could be done by making the default condition be nan.

smichr commented:
- [ ] Find out why this is failing:

```
 File "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/sympy/logic/tests/test_boolalg.py", line 800, in test_simplify_Boolean

assert simplify((x < oo) & (S(2) < x)) == (x > 2)

```

smichr commented:
@asmeurer , if you can look at the OP to see if this is all in line with what you expect, that would be helpful.

smichr committed:
add as_Basic to basic.py

smichr committed:
PEP8 edits

smichr committed:
func_name recognizes <class '

smichr committed:
no interval solutions from solve; more care with Eq/Ne

smichr committed:
! operator added to rcode; tests expanded

smichr committed:
ITE pretty prints like Piecewise

smichr committed:
modify relational simplification test

smichr committed:
ITE/Boolean/Piecewise modifications

smichr committed:
periodicity: simplify after converting Relational

smichr committed:
function_range watches for constant function

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -284,9 +363,71 @@ class BooleanFunction(Application, Boolean):
     is_Boolean = True
+    def __lt__(self, other):
+        from sympy.logic.boolalg import Boolean
+        if isinstance(self, Boolean):
```
This will always be True. BooleanFunction is a subclass of Boolean.

smichr replied:
Corrected and tested.

smichr commented on sympy/calculus/tests/test_util.py:
```diff
@@ -12,16 +12,30 @@
+        ) == Interval(-oo, -1)
+    assert function_range(sqrt(3*x - 1), x, Interval(0, 2)
+        ) == Interval(0, sqrt(5))
+    assert function_range(x*(x - 1) - (x**2 - x), x, S.Reals
```
added this and the next one -- the others are just line wrapped

smichr committed:
define default iter_solutions

smichr commented on sympy/core/basic.py:
```diff
@@ -13,6 +13,22 @@
+    """Return expr as a Basic instance using strict sympify
+    or raise a TypeError"""
+    from sympy.utilities.misc import func_name
+    if not isinstance(expr, Basic):
```
make this a simpler wrapper

smichr committed:
func_name gets short option

smichr committed:
Relationals get a count_op of 1

This allows a canonical result to be returned for Relationals
unless it is suppressed with a small ratio.

smichr committed:
remove what is already checked after #13391

smichr committed:
add note about args being generator

smichr committed:
docstring edit

smichr committed:
Py < 3: do not allow Boolean args in relationals

smichr committed:
major docstring edits

smichr committed:
add note about why evaluate=True is used

smichr committed:
Ne(x, 1).as_set().as_relational(x) -> still has Ne

smichr committed:
rewrite inequality test

smichr committed:
as_set modifications

smichr committed:
add docstring examples to as_expr_set_pairs

smichr commented on sympy/core/relational.py:
```diff
@@ -230,32 +230,14 @@ def __nonzero__(self):
-        """
+    def _eval_as_set(self):
+        # self is univariate and periodicity(self, x) in (0, None)
+        from sympy.solvers.solveset import solveset
```
remove this import

smichr committed:
update assume.py docstrings; test_args

smichr commented:
Add a test which, when it fair, will indicate that the 2.7 code can be removed.

smichr committed:
Update relational.py
